// @ledgerhq/hw-transport-node-hid-singleton/lib/TransportNodeHid.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_hid_1 = __importDefault(require("node-hid"));
const hw_transport_node_hid_noevents_1 = __importStar(require("@ledgerhq/hw-transport-node-hid-noevents"));
const logs_1 = require("@ledgerhq/logs");
const devices_1 = require("@ledgerhq/devices");
const errors_1 = require("@ledgerhq/errors");
const listenDevices_1 = require("./listenDevices");
let transportInstance;
const DISCONNECT_TIMEOUT = 5000;
let disconnectTimeout;
const clearDisconnectTimeout = () => {
    if (disconnectTimeout) {
        clearTimeout(disconnectTimeout);
    }
};
const setDisconnectTimeout = () => {
    clearDisconnectTimeout();
    disconnectTimeout = setTimeout(() => TransportNodeHidSingleton.autoDisconnect(), DISCONNECT_TIMEOUT);
};
/**
 * node-hid Transport implementation
 * @example
 * import TransportNodeHid from "@ledgerhq/hw-transport-node-hid-singleton";
 * ...
 * TransportNodeHid.create().then(transport => ...)
 */
class TransportNodeHidSingleton extends hw_transport_node_hid_noevents_1.default {
    constructor() {
        super(...arguments);
        this.preventAutoDisconnect = false;
    }
    /**
     * convenience wrapper for auto-disconnect logic
     */
    static autoDisconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            if (transportInstance && !transportInstance.preventAutoDisconnect) {
                (0, logs_1.log)("hid-verbose", "triggering auto disconnect");
                TransportNodeHidSingleton.disconnect();
            }
            else if (transportInstance) {
                // If we have disabled the auto-disconnect, try again in DISCONNECT_TIMEOUT
                clearDisconnectTimeout();
                setDisconnectTimeout();
            }
        });
    }
    /**
     * globally disconnect the transport singleton
     */
    static disconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            if (transportInstance) {
                transportInstance.device.close();
                transportInstance.emit("disconnect");
                transportInstance = null;
            }
            clearDisconnectTimeout();
        });
    }
    /**
     * if path="" is not provided, the library will take the first device
     */
    static open() {
        clearDisconnectTimeout();
        return Promise.resolve().then(() => {
            if (transportInstance) {
                (0, logs_1.log)("hid-verbose", "reusing opened transport instance");
                return transportInstance;
            }
            const device = (0, hw_transport_node_hid_noevents_1.getDevices)()[0];
            if (!device)
                throw new errors_1.CantOpenDevice("no device found");
            (0, logs_1.log)("hid-verbose", "new HID transport");
            transportInstance = new TransportNodeHidSingleton(new node_hid_1.default.HID(device.path));
            const unlisten = (0, listenDevices_1.listenDevices)(() => { }, () => {
                // assume any ledger disconnection concerns current transport
                if (transportInstance) {
                    transportInstance.emit("disconnect");
                }
            });
            const onDisconnect = () => {
                if (!transportInstance)
                    return;
                (0, logs_1.log)("hid-verbose", "transport instance was disconnected");
                transportInstance.off("disconnect", onDisconnect);
                transportInstance = null;
                unlisten();
            };
            transportInstance.on("disconnect", onDisconnect);
            return transportInstance;
        });
    }
    setAllowAutoDisconnect(allow) {
        this.preventAutoDisconnect = !allow;
    }
    /**
     * Exchange with the device using APDU protocol.
     * @param apdu
     * @returns a promise of apdu response
     */
    exchange(apdu) {
        const _super = Object.create(null, {
            exchange: { get: () => super.exchange }
        });
        return __awaiter(this, void 0, void 0, function* () {
            clearDisconnectTimeout();
            const result = yield _super.exchange.call(this, apdu);
            setDisconnectTimeout();
            return result;
        });
    }
    close() {
        // intentionally, a close will not effectively close the hid connection but
        // will allow an auto-disconnection after some inactivity
        this.preventAutoDisconnect = false;
        return Promise.resolve();
    }
}
exports.default = TransportNodeHidSingleton;
/**
 *
 */
TransportNodeHidSingleton.isSupported = hw_transport_node_hid_noevents_1.default.isSupported;
/**
 *
 */
TransportNodeHidSingleton.list = hw_transport_node_hid_noevents_1.default.list;
/**
 */
TransportNodeHidSingleton.listen = (observer) => {
    let unsubscribed;
    Promise.resolve((0, hw_transport_node_hid_noevents_1.getDevices)()).then((devices) => {
        // this needs to run asynchronously so the subscription is defined during this phase
        for (const device of devices) {
            if (!unsubscribed) {
                const deviceModel = (0, devices_1.identifyUSBProductId)(device.productId);
                observer.next({
                    type: "add",
                    descriptor: "",
                    device: {
                        name: device.deviceName,
                    },
                    deviceModel,
                });
            }
        }
    });
    const onAdd = (device) => {
        const deviceModel = (0, devices_1.identifyUSBProductId)(device.productId);
        observer.next({
            type: "add",
            descriptor: "",
            deviceModel,
            device: {
                name: device.deviceName,
            },
        });
    };
    const onRemove = (device) => {
        const deviceModel = (0, devices_1.identifyUSBProductId)(device.productId);
        observer.next({
            type: "remove",
            descriptor: "",
            deviceModel,
            device: {
                name: device.deviceName,
            },
        });
    };
    const stop = (0, listenDevices_1.listenDevices)(onAdd, onRemove);
    function unsubscribe() {
        stop();
        unsubscribed = true;
    }
    return {
        unsubscribe,
    };
};
